### PR TITLE
Comet 0.15.0 blog post

### PR DESCRIPTION
**Pull request by andygrove**
Staged version: https://datafusion.staged.apache.org/blog/2026/04/18/datafusion-comet-0.15.0/

### CODE REVIEW

File: content/blog/2026-04-18-datafusion-comet-0.15.0.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+## Other Key Features
+
+### New Expressions and Function Support
+
+This release adds support for the following:
+
+- **Date/time functions**: `days`, `hours`, `date_from_unix_date`
+- **String/JSON functions**: native `get_json_object` with improved performance over the fallback path
+- **Hash/math functions**: `crc32c`, `bin`
```
Comment (mbutrovich) [inline content/blog/2026-04-18-datafusion-comet-0.15.0.md]: Just to clarify: `crc32c` wasn't an added expression. It's a new checksum algorithm for Spark 4.0 for shuffle files.

File: content/blog/2026-04-18-datafusion-comet-0.15.0.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  continuous validation of the native Iceberg code path. Test diffs for Spark 3.4 were updated to keep the matrix
+  green across supported Spark versions.
+- **iceberg-rust upgrade**: Comet picks up the latest iceberg-rust, pulling in fixes for Parquet reader edge cases
+  discovered in earlier testing.
+- **Refreshed documentation**: The Iceberg user guide has been rewritten to reflect current capabilities, and the
+  contributor guide now documents how to run the Iceberg Spark test suites locally.
+
+Users who need to fall back to the previous behavior can still opt out, but we encourage the community to exercise
+the native reader and report any issues.
+
```
Comment (mbutrovich) [inline content/blog/2026-04-18-datafusion-comet-0.15.0.md:130]:
```suggestion
### Sort-Merge Join Performance

Comet relies heavily on sort-merge join (SMJ) because DataFusion's hash joins do not yet support spilling to
disk. For larger-than-memory joins, SMJ is the only viable path, making its performance critical for real-world
workloads at scale.

DataFusion 53 includes several SMJ improvements that Comet 0.15.0 benefits from directly:

- **Zero-copy slicing** instead of the take kernel ([datafusion#20463](https://github.com/apache/datafusion/pull/20463))
- **Streaming output** instead of waiting for all input before emitting ([datafusion#20482](https://github.com/apache/datafusion/pull/20482))
- **Cached row counts** to avoid O(n) recounting ([datafusion#20478](https://github.com/apache/datafusion/pull/20478))

Additional SMJ work is landing in upstream DataFusion and will arrive in a future Comet release:

- Specialized semi/anti join stream ([datafusion#20806](https://github.com/apache/datafusion/pull/20806))
- Batch deferred filtering with 20–50x improvements for near-unique LEFT and FULL joins ([datafusion#21184](https://github.com/apache/datafusion/pull/21184))
- DynComparator for ~5% TPC-H improvement ([datafusion#21484](https://github.com/apache/datafusion/pull/21484))
- Vec-based filter state replacing HashMap ([datafusion#21517](https://github.com/apache/datafusion/pull/21517))
- Full outer join correctness fix for NULL filter results ([datafusion#21660](https://github.com/apache/datafusion/pull/21660))

With these performance improvements, the next release of Comet will enable SMJ with filters by default.
```

File: content/blog/2026-04-18-datafusion-comet-0.15.0.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+Together, these changes reduce CPU and memory overhead for shuffle-heavy, broadcast-heavy, and
+object-storage-bound workloads.
+
+## Native Iceberg Reader Enabled by Default
+
+This release marks a major milestone for Iceberg users: **Comet's fully-native Iceberg reader is now enabled by
+default**. Workloads that read Iceberg tables will automatically benefit from native Rust-based scans built on
+iceberg-rust, with no additional configuration required.
+
+To support this change, the release bundles a broad set of Iceberg-focused improvements:
```
Comment (kazuyukitanimura) [inline content/blog/2026-04-18-datafusion-comet-0.15.0.md:112]: One of the key advantages of using the native iceberg reader is DPP support. Should we call out?

**Review by kazuyukitanimura — COMMENTED** [review]
Overall looks good.

File: content/blog/2026-04-18-datafusion-comet-0.15.0.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+- **Object store caching**: Object stores and bucket region lookups are cached, dramatically reducing DNS query
+  volume on workloads that open many files.
+- **`get_ranges` performance**: Picked up an upstream `opendal` fix that restores fast range reads from object
+  storage.
+
+Together, these changes reduce CPU and memory overhead for shuffle-heavy, broadcast-heavy, and
+object-storage-bound workloads.
+
+## Native Iceberg Reader Enabled by Default
```
Comment (kevinjqliu) [inline content/blog/2026-04-18-datafusion-comet-0.15.0.md:106]: 🥳

File: content/blog/2026-04-18-datafusion-comet-0.15.0.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+- **Trigonometric compatibility**: `tan` and `atan2` are now Spark-compatible.
+
+## Dependency Upgrades
+
+This release upgrades to **DataFusion 53.1** and **Arrow 58.1**, and picks up the latest `iceberg-rust` release
+with additional reader fixes. The `jni` crate was upgraded to 0.22.4.
+
+## Deprecations and Removals
+
+The `SupportsComet` interface has been removed, along with the Java-based Iceberg integration path (which is
```
Comment (kevinjqliu) [inline content/blog/2026-04-18-datafusion-comet-0.15.0.md:200]: do you think this is a good place to link to https://github.com/apache/datafusion-comet/issues/2921?

**Review by kevinjqliu — APPROVED** [review]
LGTM, great writeup

hyperlinks all work

File: content/blog/2026-04-18-datafusion-comet-0.15.0.md (hunk 5 above)
Comment (andygrove) [reply content/blog/2026-04-18-datafusion-comet-0.15.0.md:200]: Added, thanks

File: content/blog/2026-04-18-datafusion-comet-0.15.0.md (hunk 3 above)
Comment (andygrove) [reply content/blog/2026-04-18-datafusion-comet-0.15.0.md:112]: Added, thanks

File: content/blog/2026-04-18-datafusion-comet-0.15.0.md (hunk 1 above)
Comment (andygrove) [reply content/blog/2026-04-18-datafusion-comet-0.15.0.md]: Removed, thanks

**Review by kazuyukitanimura — APPROVED** [review]

**Review by mbutrovich — APPROVED** [review]
Thanks @andygrove!

**Comment by andygrove** [comment]
Merged. Thank you @kazuyukitanimura @mbutrovich @kevinjqliu!
